test/versioned/elastic/elasticsearch.tap.js

```diff
@@ -0,0 +1,385 @@
+ */
+
+'use strict'
+
+const tap = require('tap')
+const test = tap.test
+const helper = require('../../lib/agent_helper')
+const params = require('../../lib/params')
+const urltils = require('../../../lib/util/urltils')
+
+const dbTools = (client) => {
+  const createIndex = async (index) => {
+    return await client.indices.create({
+      index
+    })
+  }
+
+  const indexExists = async (index) => {
+    return await client.indices.exists({
+      index
+    })
+  }
+
+  const documentExists = async (index, id) => {
+    return await client.exists({
+      id,
+      index
+    })
+  }
+
+  const searchDocument = async (index, title) => {
+    return await client.search({
+      index,
+      query: { fuzzy: { title } }
+    })
+  }
+
+  const createDocument = async (index, id, document) => {
+    return await client.index({
+      index,
+      id,
+      document
+    })
+  }
+
+  const deleteDocument = async (index, id) => {
+    return await client.delete({
+      id,
+      index
+    })
+  }
+
+  return {
+    createIndex,
+    indexExists,
+    documentExists,
+    searchDocument,
+    createDocument,
+    deleteDocument
+  }
+}
+
+test('Elasticsearch instrumentation', { timeout: 20000 }, (t) => {
+  t.autoend()
+
+  let METRIC_HOST_NAME = null
+  let HOST_ID = null
+  const DB_INDEX = `test`
+  const DB_INDEX_2 = `test2`
+
+  let agent
+  let client
+  let db
+
+  t.beforeEach(async () => {
+    agent = helper.instrumentMockedAgent()
+
+    METRIC_HOST_NAME = urltils.isLocalhost(params.elastic_host)
+      ? agent.config.getHostnameSafe()
+      : params.elastic_host
+    HOST_ID = METRIC_HOST_NAME + '/' + params.elastic_port
+
+    // need to capture attributes
+    agent.config.attributes.enabled = true
+
+    const { Client, HttpConnection } = require('@elastic/elasticsearch')
+    client = new Client({
+      node: `http://${params.elastic_host}:${params.elastic_port}`,
+      auth: {
+        username: params.elastic_user,
+        password: params.elastic_pass
+      },
+      Connection: HttpConnection
+    })
+    if (!db) {
+      db = dbTools(client)
+    }
+
+    // set up index
+    const hasIndex = await db.indexExists(DB_INDEX)
+    if (!hasIndex) {
+      await db.createIndex(DB_INDEX)
+    }
+
+    // Start testing!
+    // t.notOk(agent.getTransaction(), 'no transaction should be in play')
+  })
+
+  t.afterEach(async () => {
+    // we may have to purge require cache of redis related instrumentation
+    // otherwise it will not re-register on subsequent test runs
+    // Object.keys(require.cache).forEach((key) => {
+    //   if (/elastic/.test(key)) {
+    //     delete require.cache[key]
+    //   }
+    // })
+
+    agent && helper.unloadAgent(agent)
+  })
+
+  t.teardown(async () => {
+    await client.indices.delete({ index: DB_INDEX })
+    await client.indices.delete({ index: DB_INDEX_2 })
+  })
+
+  t.test('should find Elastic calls in the transaction trace', async function (t) {
```

Comment:
most of these tests is around indices.  The index operations are relatively consistent regarding APM. Complexities of searching is what's worth testing.  That's why the instrumentation is showing uncovered in a couple of places.